File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        if(root == NULL) return 0;
        queue<pair<TreeNode*, int>> q;
        int cnt = 0;
        q.push({root, INT_MIN});
        while(!q.empty()){
            auto& [node, maxVal] = q.front();
            q.pop();
            if(node->val >= maxVal) cnt++;
            int newVal = max(maxVal, node->val);
            if(node->left) q.push({node->left, newVal});
            if(node->right) q.push({node->right, newVal});
        }
        return cnt;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.k = k
        if not root:
            return -1
        left = self.kthSmallest(root.left, self.k)    
        if left != -1:
            return left
        self.k -= 1
        if self.k == 0:
            return root.val
        return self.kthSmallest(root.right, self.k)

